### ISSUE TITLE
Text/IText _initDimensions should also clear _charWidthsCache

### ISSUE THREAD

**Issue by gordyr** [kangax/fabric.js#2018]
I've just noticed this and since it is only a single line of code it is not worth me forking and creating a PR.

It is however an important fix.

When we perform a ```set``` command on any ```Text/IText``` we check to see if any of the properties effect the objects dimensions, if so we fire off another call to ```_initDimensions```.

```_initDimensions``` in turns fires off ```_clearCache``` which clears the cache for all line widths, heights etc then recalculates them all.

This works perfectly almost 100% of the time, however when changing the actual font family itself, the ```_charWidthsCache``` does not get cleared. Meaning that we are using outdated character widths for all subsequent operations.  Obviously this needs fixing.

**But....**

If we clear the ```_charWidthsCache``` inside the ```_clearCache``` method then we see a performance degradation since we then have to recalculate the character widths everytime the user types something.

**The solution...**

Therefore the solution is to clear ```_charWidthsCache``` inside the ```_initDimensions``` method itself, separately from the ```_clearCache``` method so that it only ever gets called when we are setting dimension effecting properties on the ```Text/Itext``` object.

Like so:

```this._charWidthsCache = [  ];```

This makes the new ```_initDimensions``` method as follows:

```
/**
         * Renders text object on offscreen canvas, so that it would get dimensions
         * @private
         */
        _initDimensions: function() {
            if (this.__skipDimension) {
                return;
            }
            this._clearCache();
            this._charWidthsCache = [ ]; //Also clear character width cache in case we have changed the font Family
            var ctx = fabric.util.createCanvasElement().getContext('2d');
            this._textLines = this.text.split(this._reNewline);
            this._setTextStyles(ctx);
            this.width = this._getTextWidth(ctx);
            this.height = this._getTextHeight(ctx);
        }
```

I noticed this while running in to troubles using custom webfonts combined with ```loadFromJSON``` the above single line fixes the issue.

**Comment by asturur** [comment]
I write down a notice here, of another thing that i need to fix about the iText, that is that fontAlign changes need some cache clearing also.
Togheter with deregister mouse:move event on exit edit.

**Comment by gordyr** [comment]
Ahhh i've just realised that this shouldn't be necessary....

 Cached character widths are stored by a key name based upon their ```fontFamily```, ```fontWeight```, and ```fontStyle```, therefore the object should never need clearing during normal use.

However, without this hack a problem is still caused when loading webFonts from JSON using ```loadFromJSON```.  This is actually down to a browser bug when getting character widths.

I don't have time to explain now but i will do so tomorrow....

In a nutshell, to fix it requires us ensuring the fonts are completely loaded before populating ```_charWidthsCache ``` for any given font. 

Given the fact that new browser now have a native font loading API (Chrome and firefox) I believe that using this would be the best solution possible.  There is a tiny polyfill available for older browsers.

I'll knock a PR up soon.

**Comment by asturur** [comment]
yes the "onfontloaded" is a long standing problem to solve. we do not know when a font has finished loading.
